titan.1
.\" Copyright (c) 2023 Omar Polo <op@omarpolo.com>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.Dd October 20, 2023
.Dt TITAN 1
.Os
.Sh NAME
.Nm titan
.Nd Titan client
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl C Ar cert
.Op Fl K Ar key
.Op Fl m Ar mime
.Op Fl t Ar token
.Ar url
.Op Ar file
.Ek
.Sh DESCRIPTION
.Nm
allows one to upload data to a Titan server.
If no
.Ar file
is given on the command line,
.Nm
reads from standard input.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl C Ar cert
Specify the TLS client certificate to use.
.It Fl K Ar cert
Specify the TLS client certificate key to use.
Defaults to
.Fl C
if provided.
.It Fl m Ar mime
Specify the mime of the content being sent.
Unset by default.
.It Fl t Ar token
Specify the token for the transaction.
Unset by default.
.El
.Pp
.Nm
alters the passed
.Ar url
to include the parameter for the file size as well as the MIME and the
token if
.Fl m
or
.Fl t
are given.
.Sh EXIT STATUS
The
.Nm
utility exits with one of the following values:
.Pp
.Bl -tag -width Ds -offset indent -compact
.It 0
The transaction completed successfully and the response code was in the
2x or 3x range.
.It 1
An error occurred.
.It 2
The response code was not in the 2x or 3x range.
.El
.Sh SEE ALSO
.Xr ftp 1 ,
.Xr gg 1
.Sh STANDARDS
.Nm
implements the
.Dq Titan Specification
.Lk gemini://transjovian.org/titan/page/The%20Titan%20Specification
.Sh AUTHORS
.An -nosplit
The
.Nm
utility was written by
.An Omar Polo Aq Mt op@omarpolo.com .
.Sh CAVEATS
.Nm
doesn't perform TOFU
.Pq Trust On First Use
or any X.509 certificate validation beyond the name verification.


ChangeLog
2024-08-23  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.1.1

2024-08-23  Omar Polo  <op@omarpolo.com>

	* work around comically tiny HOST_NAME_MAX on glibc-based systems.

2024-08-18  Christoph Liebender  <cristoph.liebender@posteo.de>

	* fcgi.c (fcgi_req): send the port, not the hostname, as SERVER_PORT.

2024-08-05  Omar Polo  <op@omarpolo.com>

	* configure: fix getprogname() check.

2024-08-03  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.1 "From Mars To Sirius"
	* add support for the proxy-protocol v1 in outgoing connections too;
	needs to be enabled with the new `proxy-v1' flag in a `proxy' block.
	* gg.c (parse_proxy): support IPv6 addresses in -P

2024-06-17  Christoph Liebender  <cristoph.liebender@posteo.de>

	* add support for the proxy-protocol v1 on incoming connection;
	needs to be enabled with the new `proxy-v1' flag for `listen'.

2024-06-17  Omar Polo  <op@omarpolo.com>

	* titan.c (open_input_file): improve error checking; errors like
	"no space left on device" would have been hidden otherwise.

2024-06-14  Omar Polo  <op@omarpolo.com>

	* gmid.c: correctly propagate the prefork value; was causing
	failures if the `prefork' setting was changed from the default.

2024-06-12  Omar Polo  <op@omarpolo.com>

	* gmid.c (main): fix typo in a bounds check resulting in
	"chroot path too long" on some systems.

2024-06-11  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0.5

2024-06-10  Omar Polo  <op@omarpolo.com>

	* don't error on a '..' component at the start of the path
	* reject NUL bytes embedded in the request

2024-06-09  Omar Polo  <op@omarpolo.com>

	* check for truncation various strlcpy calls.
	* clean up of a few unused prototypes and externs.

2024-06-08  Omar Polo  <op@omarpolo.com>

	* configure: change how strnvis(3) is handled: on systems
	with the broken interface gmid will just use its built-in
	version.

2024-06-06  Omar Polo  <op@omarpolo.com>

	* parse.y: allow again empty lines at the start of the config
	* configure (VERSION): release 2.0.4
	* portability fix for system with a wrong strnvis(3)

2024-06-05  Omar Polo  <op@omarpolo.com>

	* parse.y: add a nicer error message if the removed `cgi' option
	is still used.

2024-06-04  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0.3

2024-06-03  Omar Polo  <op@omarpolo.com>

	* server.c (vhost_fastcgi): fix `fastcgi off' handling.
	Reported by Alex // nytpu

2024-05-29  Omar Polo  <op@omarpolo.com>

	* server.c (handle_handshake): relax the SNI requirement.  There are
	legitimate use-cases where SNI can't be used, like connecting to a
	raw IPv6 address.
	* gg.c (main): add -q to avoid printing "Server Says:"
	* gg.c (main): unbreak -n
	* iri.c (parse_authority): add support for raw IPv6 addresses

2024-04-04  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0.2

2024-04-03  Omar Polo  <op@omarpolo.com>

	* configure: improve function checking in the configure
	* have/landlock.c: fix landlock test
	* gmid.c (main_print_conf): fix config dumping with -nn

2024-03-03  Omar Polo  <op@omarpolo.com>

	* gmid.c: fix `log access path' with a chroot

2024-01-30  Anna “CyberTailor”

	* contrib/vim/indent/gmid.vim: fix indent

2024-01-30  Omar Polo  <op@omarpolo.com>

	* parse.y: don't make log styles reserved keywords.  Unbreaks the
	example in the manpage with `common = ...'.

2024-01-26  Omar Polo  <op@omarpolo.com>

	* parse.y: rework grammar to allow the semicolon after
	variables/macros definition and top-level options

2024-01-24  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0.1

2024-01-21  Omar Polo  <op@omarpolo.com>

	* convert gmid to the new imsg API

2024-01-14  Anna “CyberTailor”

	* configure: fix --mandir handling

2024-01-11  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): release 2.0

2024-01-09  Anna “CyberTailor”

	* contrib/vim/syntax/gmid.vim: update Vim syntax file
	* contrib/vim/ale_listers/gmid/gmid.vim: add ALE linter

2023-10-18  Omar Polo  <op@omarpolo.com>

	* ge.c (load_local_cert): generate EC certs by default, use -R to keep generating RSA ones.
	* utils.c (gencert): generate EC too
	* gg.c (get): print the response header for non-2x replies to standard error

2023-10-15  Omar Polo  <op@omarpolo.com>

	* gg.c (main): exit with the gemini response code (unless it's 2x)

2023-08-29  Omar Polo  <op@omarpolo.com>

	* ge.c (data_dir): use $XDG_DATA_HOME/gemexp instead of /gmid for the certificates.

2023-08-18  Omar Polo  <op@omarpolo.com>

	* fcgi.c (fcgi_req): sync the parameters with RFC3875 (CGI)
	(fcgi_req): send "GET" as REQUEST_METHOD

2023-08-08  Omar Polo  <op@omarpolo.com>

	* fcgi.c (fcgi_req): implement SCRIPT_NAME / PATH_INFO splitting for fastcgi
	(fcgi_req): add `fastcgi strip'

2023-08-07  Omar Polo  <op@omarpolo.com>

	* logger.c (logger_dispatch_server): allow to change the syslog(3) facility.

	* gmid.c (main): attempt to load TLS certificates and load mimes and virtual hosts root as part of configtest (-n) rather than verifying the syntax of the configuration only.

2023-07-25  Omar Polo  <op@omarpolo.com>

	* gmid.c (log_request): allow to change the logging style.

2023-07-24  Omar Polo  <op@omarpolo.com>

	* parse.y: add ability to log to files with `log access <path>'.

2023-07-23  Omar Polo  <op@omarpolo.com>

	* parse.y: revamp fastcgi configuration; make it per-location

2023-07-22  Omar Polo  <op@omarpolo.com>

	* titan.c (main): add titan(1), a simple titan client

	* gg.c (get): warn when the server doesn't use TLS' close notify

2023-07-01  Omar Polo  <op@omarpolo.com>

	* fcgi.c (fcgi_handle_stdout): parse and log the fastcgi reply

2023-06-24  Omar Polo  <op@omarpolo.com>

	* server.c (handle_handshake): correctly handle TLS handshake failures.

	* server.c (client_close_ev): plug memory leak

2023-06-23  Omar Polo  <op@omarpolo.com>

	* parse.y: implement `listen on'

2023-06-13  Omar Polo  <op@omarpolo.com>


titan.c
/*
 * Copyright (c) 2023 Omar Polo <op@omarpolo.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "config.h"

#include <sys/stat.h>
#include <sys/socket.h>

#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <poll.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <tls.h>
#include <unistd.h>

#include "iri.h"

#ifndef INFTIM
#define INFTIM -1
#endif

#ifndef __OpenBSD__
#define pledge(a, b) (0)
#endif

static int
xasprintf(char **ret, const char *fmt, ...)
{
	va_list	 ap;
	int	 r;

	va_start(ap, fmt);
	r = vasprintf(ret, fmt, ap);
	va_end(ap);

	if (r == -1)
		err(1, "asprintf");
	return r;
}

static int
dial(const char *hostname, const char *port)
{
	struct addrinfo	 hints, *res, *res0;
	int		 error, save_errno, s;
	const char	*cause = NULL;

	if (port == NULL || *port == '\0')
		port = "1965";

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	error = getaddrinfo(hostname, port, &hints, &res0);
	if (error)
		errx(1, "can't resolve %s: %s", hostname, gai_strerror(error));

	s = -1;
	for (res = res0; res; res = res->ai_next) {
		s = socket(res->ai_family, res->ai_socktype,
		    res->ai_protocol);
		if (s == -1) {
			cause = "socket";
			continue;
		}

		if (connect(s, res->ai_addr, res->ai_addrlen) == -1) {
			cause = "connect";
			save_errno = errno;
			close(s);
			errno = save_errno;
			s = -1;
			continue;
		}

		if (fcntl(s, F_SETFL, O_NONBLOCK) == -1)
			err(1, "fcntl");
		break; /* got one */
	}
	if (s == -1)
		err(1, "%s", cause);
	freeaddrinfo(res0);
	return (s);
}

/* returns read bytes, or -1 on error */
static ssize_t
iomux(struct tls *ctx, int fd, const char *in, size_t inlen, char *out,
    size_t outlen)
{
	struct pollfd	 pfd;
	size_t		 outwrote = 0;
	ssize_t		 ret;

	memset(&pfd, 0, sizeof(pfd));
	pfd.fd = fd;
	pfd.events = POLLIN|POLLOUT;

	for (;;) {
		if (poll(&pfd, 1, INFTIM) == -1)
			err(1, "poll");
		if (pfd.revents & (POLLERR|POLLNVAL))
			errx(1, "bad fd %d", pfd.fd);

		/* attempt to read */
		if (out != NULL) {
			switch (ret = tls_read(ctx, out, outlen)) {
			case TLS_WANT_POLLIN:
			case TLS_WANT_POLLOUT:
				break;
			case -1:
				return -1;
			case 0:
				if (outwrote == 0)
					return -1;
				return outwrote;
			default:
				outwrote += ret;
				out += ret;
				outlen -= ret;
			}

			/*
			 * don't write if we're reading; titan works
			 * like this.
			 */
			if (outwrote != 0)
				continue;
		}

		if (inlen == 0 && out == NULL)
			break;
		if (inlen == 0)
			continue;

		switch (ret = tls_write(ctx, in, inlen)) {
		case TLS_WANT_POLLIN:
		case TLS_WANT_POLLOUT:
			continue;
		case 0:
		case -1:
			return -1;
		default:
			in += ret;
			inlen -= ret;
		}
	}

	return 0;
}

static FILE *
open_input_file(int argc, char **argv)
{
	FILE	*fp;
	char	 buf[BUFSIZ];
	char	 sfn[22];
	size_t	 r;
	int	 fd;

	if (argc > 1) {
		if ((fp = fopen(argv[1], "r")) == NULL)
			err(1, "can't open %s", argv[1]);
		return fp;
	}

	strlcpy(sfn, "/tmp/titan.XXXXXXXXXX", sizeof(sfn));
	if ((fd = mkstemp(sfn)) == -1 ||
	    (fp = fdopen(fd, "w+")) == NULL) {
		warn("%s", sfn);
		if (fd != -1) {
			unlink(sfn);
			close(fd);
		}
		errx(1, "can't create temp file");
	}
	unlink(sfn);

	for (;;) {
		r = fread(buf, 1, sizeof(buf), stdin);
		if (r == 0)
			break;
		if (fwrite(buf, 1, r, fp) != r)
			break;
	}
	if (ferror(fp) || ferror(stdin))
		err(1, "I/O error");

	if (fseeko(fp, 0, SEEK_SET) == -1)
		err(1, "fseeko");

	return fp;
}

static int
parse_response(char *r)
{
	int code;

	if (r[0] < '0' || r[0] > '9' ||
	    r[1] < '0' || r[1] > '9' ||
	    r[2] != ' ')
		errx(1, "illegal response");

	code = (r[0] - '0') * 10 + (r[1] - '0');
	if (code < 10 || code >= 70)
		errx(1, "invalid response code: %d", code);
	if (code >= 20 && code < 30)
		return 0;
	if (code >= 30 && code < 40) {
		puts(r + 3);
		return 0;
	}
	warnx("server error: %s", r + 3);
	return 2;
}

static void __dead
usage(void)
{
	fprintf(stderr,
	    "usage: %s [-C cert] [-K key] [-m mime] [-t token] url [file]\n",
	    getprogname());
	exit(1);
}

int
main(int argc, char **argv)
{
	struct tls_config *config;
	struct tls	*ctx;
	struct stat	 sb;
	struct iri	 iri;
	FILE		*in;
	const char	*cert = NULL, *key = NULL, *mime = NULL, *token = NULL;
	const char	*errstr;
	char		 iribuf[1025];
	char		 reqbuf[1025];
	char		 resbuf[1025];
	char		*path;
	int		 sock, ch, ret = 0;

	if (pledge("stdio rpath tmppath inet dns", NULL) == -1)
		err(1, "pledge");

	while ((ch = getopt(argc, argv, "C:K:m:t:")) != -1) {
		switch (ch) {
		case 'C':
			cert = optarg;
			break;
		case 'K':
			key = optarg;
			break;
		case 'm':
			mime = optarg;
			break;
		case 't':
			token = optarg;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (cert == NULL && key != NULL)
		usage();
	if (cert != NULL && key == NULL)
		key = cert;

	if (argc != 1 && argc != 2)
		usage();

	in = open_input_file(argc, argv);

	/* drop rpath tmppath */
	if (pledge("stdio inet dns", NULL) == -1)
		err(1, "pledge");

	if (fstat(fileno(in), &sb) == -1)
		err(1, "fstat");

	/* prepare the IRI */
	if (strlcpy(iribuf, argv[0], sizeof(iribuf)) >= sizeof(iribuf))
		errx(1, "IRI too long");

	if (!parse_iri(iribuf, &iri, &errstr))
		errx(1, "invalid IRI: %s", errstr);

	if (strcmp(iri.schema, "titan") != 0)
		errx(1, "not a titan:// IRI");

	if (token && mime) {
		xasprintf(&path, "%s;size=%lld;token=%s;mime=%s", iri.path,
		    (long long)sb.st_size, token, mime);
	} else if (token) {
		xasprintf(&path, "%s;size=%lld;token=%s", iri.path,
		    (long long)sb.st_size, token);
	} else if (mime) {
		xasprintf(&path, "%s;size=%lld;mime=%s", iri.path,
		    (long long)sb.st_size, mime);
	} else {
		xasprintf(&path, "%s;size=%lld", iri.path,
		    (long long)sb.st_size);
	}

	iri.path = path;
	if (!serialize_iri(&iri, reqbuf, sizeof(reqbuf)) ||
	    strlcat(reqbuf, "\r\n", sizeof(reqbuf)) >= sizeof(reqbuf))
		errx(1, "IRI too long");

	if ((config = tls_config_new()) == NULL)
		err(1, "tls_config_new");
	tls_config_insecure_noverifycert(config);
	tls_config_insecure_noverifyname(config);

	if (cert && tls_config_set_keypair_file(config, cert, key) == -1)
		errx(1, "can't load certificate client %s", cert);

	if ((ctx = tls_client()) == NULL)
		errx(1, "can't create tls context");

	if (tls_configure(ctx, config) == -1)
		errx(1, "tls_configure: %s", tls_error(ctx));

	sock = dial(iri.host, iri.port);

	/* drop inet tls */
	if (pledge("stdio", NULL) == -1)
		err(1, "pledge");

	if (tls_connect_socket(ctx, sock, iri.host) == -1)
		errx(1, "failed to connect to %s:%s: %s", iri.host,
		    *iri.port == '\0' ? "1965" : iri.port, tls_error(ctx));

	/* send request */
	if (iomux(ctx, sock, reqbuf, strlen(reqbuf), NULL, 0) == -1)
		errx(1, "I/O error: %s", tls_error(ctx));

	for (;;) {
		static char buf[BUFSIZ];
		size_t buflen;
		ssize_t w;
		char *m;

		/* will be zero on EOF */
		buflen = fread(buf, 1, sizeof(buf), in);

		w = iomux(ctx, sock, buf, buflen, resbuf, sizeof(resbuf));
		if (w == -1) {
			errstr = tls_error(ctx);
			if (errstr == NULL)
				errstr = "unexpected EOF";
			errx(1, "I/O error: %s", errstr);
		}
		if (w != 0) {
			if ((m = memmem(resbuf, w, "\r\n", 2)) == NULL)
				errx(1, "invalid reply");
			*m = '\0';
			ret = parse_response(resbuf);
			break;
		}
	}

	/* close connection */
	for (;;) {
		struct pollfd pfd;

		memset(&pfd, 0, sizeof(pfd));
		pfd.fd = sock;
		pfd.events = POLLIN|POLLOUT;

		switch (tls_close(ctx)) {
		case TLS_WANT_POLLIN:
		case TLS_WANT_POLLOUT:
			if (poll(&pfd, 1, INFTIM) == -1)
				err(1, "poll");
			break;
		case -1:
			warnx("tls_close: %s", tls_error(ctx));
			/* fallthrough */
		default:
			tls_free(ctx);
			return (ret);
		}
	}
}
